File: 20.valid-parentheses.py
#
# @lc app=leetcode id=20 lang=python3
#
# [20] Valid Parentheses
#

# @lc code=start
class Solution:
    def isValid(self, s: str) -> bool:
        len_s = len(s)
        if len_s % 2 ==1:
            print("1111")
            return False

        tmp = []
        
        for i in range(len_s):
            if s[i] == "(" or s[i] == "{" or s[i] == "[":
                tmp.append(s[i])
            

            if s[i] == ")" or s[i] == "}" or s[i] == "]":
                if tmp ==[]:
                    return False
                
                if s[i] == ")":
                    print(tmp[-1])
                    if tmp[-1] == "(":
                        tmp.pop(-1)
                    else:
                        break
                elif s[i] == "]":
                    if tmp[-1] == "[":
                        tmp.pop(-1)
                    else:
                        break
                elif  s[i] =="}":
                    if tmp[-1] == "{":
                        tmp.pop(-1)
                    else:
                        break
        if tmp==[]:
            return True
        return False
        
# @lc code=end

File: 20.valid-parentheses.java
/*
 * @lc app=leetcode id=20 lang=java
 *
 * [20] Valid Parentheses
 */

// @lc code=start
class Solution {
    public boolean isValid(String s) {
        Map<Character, Character> mp = new HashMap<>();
        mp.put('(',')');
        mp.put('{','}');
        mp.put('[',']');


        Stack<Character> stack = new Stack<Character>();
        
        int m = s.length();
        for(int i =0;i < m ; i++ ){
            char res = s.charAt(i);
            if(mp.containsKey(res)){
                stack.push(res);
            }
            else if (stack.isEmpty()){
                return false;
            }else{
                char tmp = stack.pop();
               // System.out.println(stack);

               if(res == mp.get(tmp)) continue;
               return false;
            }
        }
        return stack.isEmpty();
    }
}
// @lc code=end

